Assets/Tests/PlayModeTests/DeviceIdTests.cs

```diff
@@ -377,6 +372,84 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
             ValidateDeviceIDAndType(Countly.Instance, deviceID, DeviceIdType.SDKGenerated);
         }
 
+        // 'SetId' method in DeviceIdCountlyService
+        // We provide empty, null, whitespace and valid ids to SetId method over DeveloperProvided id
+        // Id should only change in case of a valid id
+        [TestCase(DeviceIdType.DeveloperProvided, "", DeviceIdType.DeveloperProvided, false)]
+        [TestCase(DeviceIdType.DeveloperProvided, null, DeviceIdType.DeveloperProvided, false)]
+        [TestCase(DeviceIdType.DeveloperProvided, " ", DeviceIdType.DeveloperProvided, false)]
+        [TestCase(DeviceIdType.DeveloperProvided, "new_user", DeviceIdType.DeveloperProvided, true)]
+        [TestCase(DeviceIdType.SDKGenerated, "", DeviceIdType.SDKGenerated, false)]
+        [TestCase(DeviceIdType.SDKGenerated, null, DeviceIdType.SDKGenerated, false)]
+        [TestCase(DeviceIdType.SDKGenerated, " ", DeviceIdType.SDKGenerated, false)]
+        [TestCase(DeviceIdType.SDKGenerated, "new_user", DeviceIdType.DeveloperProvided, true)]
+        public void SetId(DeviceIdType startingType, string id, DeviceIdType expectedIdType, bool isExpectingChange)
+        {
+            if (startingType == DeviceIdType.DeveloperProvided) {
+                Countly.Instance.Init(TestUtility.CreateBaseConfig());
+            } else {
+                Countly.Instance.Init(new CountlyConfiguration(_appKey, _serverUrl));
+            }
+
+            string deviceID = Countly.Instance.Device.DeviceId; // either becomes "test_user" or a id generated by sdk
+            ValidateDeviceIDAndType(Countly.Instance, deviceID, startingType, true);
+
+            Countly.Instance.Device.SetId(id);
+            if (isExpectingChange) {
+                ValidateDeviceIDAndType(Countly.Instance, id, expectedIdType, true);
+            } else {
+                ValidateDeviceIDAndType(Countly.Instance, deviceID, expectedIdType, true);
+            }
+        }
+
+        // 'SetId' method in DeviceIdCountlyService
+        // We provide the same device id that SDK generated 
+        // Id should remain same and type shouldn't change
+        [TestCase(DeviceIdType.SDKGenerated)]
+        [TestCase(DeviceIdType.DeveloperProvided)]
+        public void SetId_SameId(DeviceIdType startType)
+        {
+            string deviceId;
+            DeviceIdType currentType;
+            if (startType == DeviceIdType.SDKGenerated) {
```

Comment:
this code is perfectly fine. a suggestion though: you can assign a CountlyConfiguration to a configuration from the if block or the else block initially. then with a single if check you can re-assign it if necessary. this way you would eliminate the else block. only if that would be sth you also like